README.md
# csa-rs

[![Github Actions](https://github.com/nozaq/csa-rs/workflows/build/badge.svg)](https://github.com/nozaq/csa-rs/actions?workflow=build)
[![Coverage Status](https://coveralls.io/repos/github/nozaq/csa-rs/badge.svg)](https://coveralls.io/github/nozaq/csa-rs)
[![crates.io](https://img.shields.io/crates/v/csa.svg)](https://crates.io/crates/csa)
[![docs.rs](https://docs.rs/csa/badge.svg)](https://docs.rs/csa)

A Shogi game serialization/deserialization library in CSA format.
CSA format is a plaintext format for recording Shogi games. This library supports parsing CSA-formatted string as well as composing CSA-formatted string from structs. Detail about CSA format is found at [here](http://www.computer-shogi.org/protocol/record_v22.html).

[Documentation](https://docs.rs/csa)

## Usage

Below is an example of parsing CSA-formatted string into structs.

```rust
use std::time::Duration;
use csa::{parse_csa, Action, Color, GameRecord, MoveRecord, PieceType, Square};

let csa_str = "\
V2.2
N+NAKAHARA
N-YONENAGA
$EVENT:13th World Computer Shogi Championship
PI
+
+2726FU
T12
";

let game = parse_csa(csa_str).expect("failed to parse the csa content");
assert_eq!(game.black_player, Some("NAKAHARA".to_string()));
assert_eq!(game.white_player, Some("YONENAGA".to_string()));
assert_eq!(game.event, Some("13th World Computer Shogi Championship".to_string()));
assert_eq!(game.moves[0],  MoveRecord{
    action: Action::Move(Color::Black, Square::new(2, 7), Square::new(2, 6), PieceType::Pawn),
    time: Some(Duration::from_secs(12))
});
```

In contrast, structs can be composed into CSA-formatted string.

```rust
use std::time::Duration;
use csa::{ Action, Color, GameRecord, MoveRecord, PieceType, Square};

let mut g = GameRecord::default();
g.black_player = Some("NAKAHARA".to_string());
g.white_player = Some("YONENAGA".to_string());
g.event = Some("13th World Computer Shogi Championship".to_string());
g.moves.push(MoveRecord {
    action: Action::Move(
        Color::Black,
        Square::new(2, 7),
        Square::new(2, 6),
        PieceType::Pawn,
    ),
    time: Some(Duration::from_secs(5)),
});
g.moves.push(MoveRecord {
    action: Action::Toryo,
    time: None,
});

let csa_str = "\
V2.2
N+NAKAHARA
N-YONENAGA
$EVENT:13th World Computer Shogi Championship
PI
+
+2726FU
T5
%TORYO
";

assert_eq!(csa_str, g.to_string());
```

## License

`csa-rs` is licensed under the MIT license. Please read the [LICENSE](LICENSE) file in this repository for more information.


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="1.0.1"></a>
## [1.0.1] - 2022-03-16
### Fix
- update parser for move records with comments ([#18](https://github.com/nozaq/shogi-rs/issues/18))


<a name="1.0.0"></a>
## [1.0.0] - 2021-12-27
### Chore
- upgrade checkout action to v2 ([#13](https://github.com/nozaq/shogi-rs/issues/13))
- rename `master` branch to `main` ([#11](https://github.com/nozaq/shogi-rs/issues/11))


<a name="0.6.1"></a>
## [0.6.1] - 2021-08-15
### Refactor
- upgrade dependencies ([#10](https://github.com/nozaq/shogi-rs/issues/10))


<a name="0.6.0"></a>
## [0.6.0] - 2020-11-14
### Chore
- use tarpaulin to measure coverage

### Fix
- clippy warnings


<a name="0.5.0"></a>
## [0.5.0] - 2020-01-02

<a name="0.5,0"></a>
## [0.5,0] - 2020-01-02
### Feat
- upgrade to nom5
- upgrade to nom4


<a name="0.4.0"></a>
## [0.4.0] - 2020-01-01
### Chore
- use Github Actions
- use git-chglog to generate CHANGELOG

### Refactor
- cargo clippy
- cargo fmt


<a name="0.3.0"></a>
## [0.3.0] - 2019-01-27
### Fix
- add Cargo.toml

### Refactor
- migrate to 2018 edition


<a name="0.2.0"></a>
## [0.2.0] - 2017-06-24
### Chore
- remove redundant builds


<a name="0.1"></a>
## 0.1 - 2017-06-13

[Unreleased]: https://github.com/nozaq/shogi-rs/compare/1.0.1...HEAD
[1.0.1]: https://github.com/nozaq/shogi-rs/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/nozaq/shogi-rs/compare/0.6.1...1.0.0
[0.6.1]: https://github.com/nozaq/shogi-rs/compare/0.6.0...0.6.1
[0.6.0]: https://github.com/nozaq/shogi-rs/compare/0.5.0...0.6.0
[0.5.0]: https://github.com/nozaq/shogi-rs/compare/0.5,0...0.5.0
[0.5,0]: https://github.com/nozaq/shogi-rs/compare/0.4.0...0.5,0
[0.4.0]: https://github.com/nozaq/shogi-rs/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/nozaq/shogi-rs/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/nozaq/shogi-rs/compare/0.1...0.2.0


src/lib.rs
//! A serialization/deserialization library for [CSA] format.
//!
//! [CSA] format is a plaintext format for recording Shogi games.
//! This library supports parsing CSA-formatted string as well as composing CSA-formatted string from structs.
//! Detail about CSA format is found at http://www.computer-shogi.org/protocol/record_v22.html.
//!
//! # Examples
//! Below is an example of parsing CSA-formatted string into structs.
//!
//! ```
//! use std::time::Duration;
//! use csa::{parse_csa, Action, Color, GameRecord, MoveRecord, PieceType, Square};
//!
//! let csa_str = "\
//! V2.2
//! N+NAKAHARA
//! N-YONENAGA
//! $EVENT:13th World Computer Shogi Championship
//! PI
//! +
//! +2726FU
//! T12
//! ";
//!
//! let game = parse_csa(csa_str).expect("failed to parse the csa content");
//! assert_eq!(game.black_player, Some("NAKAHARA".to_string()));
//! assert_eq!(game.white_player, Some("YONENAGA".to_string()));
//! assert_eq!(game.event, Some("13th World Computer Shogi Championship".to_string()));
//! assert_eq!(game.moves[0],  MoveRecord{
//!     action: Action::Move(Color::Black, Square::new(2, 7), Square::new(2, 6), PieceType::Pawn),
//!     time: Some(Duration::from_secs(12))
//! });
//! ```
//!
//! In contrast, structs can be composed into CSA-formatted string.
//!
//! ```
//! use std::time::Duration;
//! use csa::{ Action, Color, GameRecord, MoveRecord, PieceType, Square};
//!
//! let mut g = GameRecord::default();
//! g.black_player = Some("NAKAHARA".to_string());
//! g.white_player = Some("YONENAGA".to_string());
//! g.event = Some("13th World Computer Shogi Championship".to_string());
//! g.moves.push(MoveRecord {
//!     action: Action::Move(
//!         Color::Black,
//!         Square::new(2, 7),
//!         Square::new(2, 6),
//!         PieceType::Pawn,
//!     ),
//!     time: Some(Duration::from_secs(5)),
//! });
//! g.moves.push(MoveRecord {
//!     action: Action::Toryo,
//!     time: None,
//! });
//!
//! let csa_str = "\
//! V2.2
//! N+NAKAHARA
//! N-YONENAGA
//! $EVENT:13th World Computer Shogi Championship
//! PI
//! +
//! +2726FU
//! T5
//! %TORYO
//! ";
//!
//! assert_eq!(csa_str, g.to_string());
//! ```
//!
//! [CSA]: http://www2.computer-shogi.org/protocol/record_v22.html

pub mod parser;
pub mod value;

pub use parser::*;
pub use value::*;


src/parser/mod.rs
mod game;
mod time;

use std::error::Error;
use std::fmt;

use self::game::game_record;
use crate::value::GameRecord;

#[derive(Debug)]
pub enum CsaError {
    ParseError(),
}

impl fmt::Display for CsaError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            CsaError::ParseError() => write!(f, "failed to parse"),
        }
    }
}

impl Error for CsaError {}

////////////////////////////////////////////////////////////////////////////////

pub fn parse_csa(s: &str) -> Result<GameRecord, CsaError> {
    if let Ok((_, record)) = game_record(s.as_bytes()) {
        Ok(record)
    } else {
        Err(CsaError::ParseError())
    }
}

////////////////////////////////////////////////////////////////////////////////

#[cfg(test)]
mod tests {
    use super::*;

    use std::fs::{self, File};
    use std::io::Read;
    use std::path::Path;

    #[test]
    fn load_fixtures() {
        let fixtures_dir = Path::new("fixtures/");
        let dir_entries = fs::read_dir(fixtures_dir).unwrap();

        for entry in dir_entries {
            let path = entry.unwrap().path();
            if !path.is_file() {
                continue;
            }

            let mut file = File::open(&path).unwrap();
            let mut contents = String::new();
            file.read_to_string(&mut contents)
                .expect("failed to load a fixuture content");
            let res = parse_csa(&contents);

            assert!(res.is_ok());
        }
    }
}


src/parser/time.rs
use nom::bytes::complete::{tag, take};
use nom::character::complete::digit1;
use nom::combinator::{map_res, opt, verify};
use nom::sequence::preceded;
use nom::*;
use std::convert::TryFrom;
use std::str;
use std::time::Duration;
use time::{Date as NativeDate, Month, Time as NativeTime};

use crate::value::{Time, TimeLimit};

fn take_2_digits(input: &[u8]) -> IResult<&[u8], i32> {
    map_res(map_res(take(2usize), str::from_utf8), |s: &str| s.parse())(input)
}

fn take_4_digits(input: &[u8]) -> IResult<&[u8], i32> {
    map_res(map_res(take(4usize), str::from_utf8), |s: &str| s.parse())(input)
}

fn take_n_digits(input: &[u8]) -> IResult<&[u8], i32> {
    map_res(map_res(digit1, str::from_utf8), |s: &str| s.parse())(input)
}

fn year(input: &[u8]) -> IResult<&[u8], i32> {
    take_4_digits(input)
}

fn month(input: &[u8]) -> IResult<&[u8], i32> {
    verify(take_2_digits, |&d| d > 0 && d < 13)(input)
}

fn day(input: &[u8]) -> IResult<&[u8], i32> {
    verify(take_2_digits, |&d| d > 0 && d < 32)(input)
}

fn hour(input: &[u8]) -> IResult<&[u8], i32> {
    verify(take_2_digits, |&d| (0..24).contains(&d))(input)
}

fn minutes(input: &[u8]) -> IResult<&[u8], i32> {
    verify(take_2_digits, |&d| (0..60).contains(&d))(input)
}

fn seconds(input: &[u8]) -> IResult<&[u8], i32> {
    verify(take_2_digits, |&d| (0..60).contains(&d))(input)
}

fn date(input: &[u8]) -> IResult<&[u8], NativeDate> {
    let (input, year) = year(input)?;
    let (input, _) = tag("/")(input)?;
    let (input, month) = month(input)?;
    let (input, _) = tag("/")(input)?;
    let (input, day) = day(input)?;

    let month = Month::try_from(month as u8).unwrap();
    let date = NativeDate::from_calendar_date(year, month, day as u8).unwrap();

    Ok((input, date))
}

fn time(input: &[u8]) -> IResult<&[u8], NativeTime> {
    let (input, hour) = hour(input)?;
    let (input, _) = tag(":")(input)?;
    let (input, minutes) = minutes(input)?;
    let (input, _) = tag(":")(input)?;
    let (input, seconds) = seconds(input)?;

    Ok((
        input,
        time::Time::from_hms(hour as u8, minutes as u8, seconds as u8).unwrap(),
    ))
}

pub fn datetime(input: &[u8]) -> IResult<&[u8], Time> {
    let (input, date) = date(input)?;
    let (input, time) = opt(preceded(tag(" "), time))(input)?;

    Ok((input, Time { date, time }))
}

pub fn timelimit(input: &[u8]) -> IResult<&[u8], TimeLimit> {
    let (input, hour) = take_n_digits(input)?;
    let (input, _) = tag(":")(input)?;
    let (input, minutes) = minutes(input)?;
    let (input, _) = tag("+")(input)?;
    let (input, byoyomi) = take_n_digits(input)?;

    Ok((
        input,
        TimeLimit {
            main_time: Duration::from_secs(hour as u64 * 60 * 60 + minutes as u64 * 60),
            byoyomi: Duration::from_secs(byoyomi as u64),
        },
    ))
}

////////////////////////////////////////////////////////////////////////////////

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_year() {
        assert_eq!(year(b"0001"), Result::Ok((&b""[..], 1)));
        assert_eq!(year(b"9999"), Result::Ok((&b""[..], 9999)));
    }

    #[test]
    fn parse_month() {
        assert!(month(b"00").is_err());
        assert_eq!(month(b"01"), Result::Ok((&b""[..], 1)));
        assert_eq!(month(b"12"), Result::Ok((&b""[..], 12)));
        assert!(month(b"13").is_err());
    }

    #[test]
    fn parse_day() {
        assert!(day(b"00").is_err());
        assert_eq!(day(b"01"), Result::Ok((&b""[..], 1)));
        assert_eq!(day(b"31"), Result::Ok((&b""[..], 31)));
        assert!(day(b"32").is_err());
    }

    #[test]
    fn parse_hour() {
        assert_eq!(hour(b"00"), Result::Ok((&b""[..], 0)));
        assert_eq!(hour(b"01"), Result::Ok((&b""[..], 1)));
        assert_eq!(hour(b"23"), Result::Ok((&b""[..], 23)));
        assert!(hour(b"25").is_err());
    }

    #[test]
    fn parse_minutes() {
        assert_eq!(minutes(b"00"), Result::Ok((&b""[..], 0)));
        assert_eq!(minutes(b"01"), Result::Ok((&b""[..], 1)));
        assert_eq!(minutes(b"59"), Result::Ok((&b""[..], 59)));
        assert!(minutes(b"60").is_err());
    }

    #[test]
    fn parse_seconds() {
        assert_eq!(seconds(b"00"), Result::Ok((&b""[..], 0)));
        assert_eq!(seconds(b"01"), Result::Ok((&b""[..], 1)));
        assert_eq!(seconds(b"59"), Result::Ok((&b""[..], 59)));
        assert!(seconds(b"60").is_err());
    }

    #[test]
    fn parse_date() {
        assert_eq!(
            date(b"2002/01/01"),
            Result::Ok((
                &b""[..],
                NativeDate::from_calendar_date(2002, Month::January, 1).unwrap()
            ))
        );
    }

    #[test]
    fn parse_time() {
        assert_eq!(
            time(b"19:00:00"),
            Result::Ok((&b""[..], NativeTime::from_hms(19, 0, 0).unwrap()))
        );
    }

    #[test]
    fn parse_datetime() {
        assert_eq!(
            datetime(b"2002/01/01"),
            Result::Ok((
                &b""[..],
                Time {
                    date: NativeDate::from_calendar_date(2002, Month::January, 1).unwrap(),
                    time: None
                }
            ))
        );
        assert_eq!(
            datetime(b"2002/01/01 19:00:00"),
            Result::Ok((
                &b""[..],
                Time {
                    date: NativeDate::from_calendar_date(2002, Month::January, 1).unwrap(),
                    time: Some(NativeTime::from_hms(19, 0, 0).unwrap())
                }
            ))
        );
    }

    #[test]
    fn parse_timelimit() {
        assert_eq!(
            timelimit(b"00:25+00"),
            Result::Ok((
                &b""[..],
                TimeLimit {
                    main_time: Duration::from_secs(25 * 60),
                    byoyomi: Duration::from_secs(0)
                }
            ))
        );
        assert_eq!(
            timelimit(b"00:30+30"),
            Result::Ok((
                &b""[..],
                TimeLimit {
                    main_time: Duration::from_secs(30 * 60),
                    byoyomi: Duration::from_secs(30)
                }
            ))
        );
        assert_eq!(
            timelimit(b"00:00+30"),
            Result::Ok((
                &b""[..],
                TimeLimit {
                    main_time: Duration::from_secs(0),
                    byoyomi: Duration::from_secs(30)
                }
            ))
        );
    }
}


src/value.rs
use std::fmt;
use std::time::Duration;
use time::{Date as NativeDate, Time as NativeTime};

#[derive(Default, Debug, PartialEq, Eq)]
pub struct GameRecord {
    pub black_player: Option<String>,
    pub white_player: Option<String>,
    pub event: Option<String>,
    pub site: Option<String>,
    pub start_time: Option<Time>,
    pub end_time: Option<Time>,
    pub time_limit: Option<TimeLimit>,
    pub opening: Option<String>,
    pub start_pos: Position,
    pub moves: Vec<MoveRecord>,
}

impl fmt::Display for GameRecord {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "V2.2")?;

        // Metadata
        let metadata = [
            ("N+", self.black_player.as_ref().map(|x| x.to_string())),
            ("N-", self.white_player.as_ref().map(|x| x.to_string())),
            ("$EVENT:", self.event.as_ref().map(|x| x.to_string())),
            ("$SITE:", self.site.as_ref().map(|x| x.to_string())),
            (
                "$START_TIME:",
                self.start_time.as_ref().map(|x| x.to_string()),
            ),
            ("$END_TIME:", self.end_time.as_ref().map(|x| x.to_string())),
            (
                "$TIME_LIMIT:",
                self.time_limit.as_ref().map(|x| x.to_string()),
            ),
            ("$OPENING:", self.opening.as_ref().map(|x| x.to_string())),
        ];
        for &(ref key, ref value) in &metadata {
            if let Some(ref value) = *value {
                writeln!(f, "{}{}", key, value)?;
            }
        }

        // Position
        write!(f, "{}", self.start_pos)?;

        // Move records
        for record in &self.moves {
            write!(f, "{}", record)?;
        }

        Ok(())
    }
}

////////////////////////////////////////////////////////////////////////////////

#[derive(Debug, PartialEq, Eq, Clone)]

pub struct Time {
    pub date: NativeDate,
    pub time: Option<NativeTime>,
}

impl Time {
    pub fn now() -> Self {
        let now = time::OffsetDateTime::now_utc();

        Time {
            date: now.date(),
            time: Some(now.time()),
        }
    }
}

impl fmt::Display for Time {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{}/{:02}/{:02}",
            self.date.year(),
            self.date.month() as u8,
            self.date.day()
        )?;
        if let Some(time) = self.time {
            write!(
                f,
                " {}:{:02}:{:02}",
                time.hour(),
                time.minute(),
                time.second()
            )?;
        }

        Ok(())
    }
}

////////////////////////////////////////////////////////////////////////////////

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct TimeLimit {
    pub main_time: Duration,
    pub byoyomi: Duration,
}

impl fmt::Display for TimeLimit {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let secs = self.main_time.as_secs();
        let hours = secs / 3600;
        let minutes = (secs % 3600) / 60;

        write!(
            f,
            "{:02}:{:02}+{:02}",
            hours,
            minutes,
            self.byoyomi.as_secs()
        )
    }
}

////////////////////////////////////////////////////////////////////////////////

#[derive(Debug, PartialEq, Eq)]
pub enum GameAttribute {
    Time(Time),
    TimeLimit(TimeLimit),
    Str(String),
}

impl fmt::Display for GameAttribute {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            GameAttribute::Time(ref time) => write!(f, "{}", time),
            GameAttribute::TimeLimit(ref time_limit) => write!(f, "{}", time_limit),
            GameAttribute::Str(ref s) => write!(f, "{}", s),
        }
    }
}

////////////////////////////////////////////////////////////////////////////////

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum Color {
    Black,
    White,
}

impl Default for Color {
    fn default() -> Self {
        Color::Black
    }
}

impl fmt::Display for Color {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            Color::Black => write!(f, "+"),
            Color::White => write!(f, "-"),
        }
    }
}

////////////////////////////////////////////////////////////////////////////////

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub struct Square {
    pub file: u8,
    pub rank: u8,
}

impl Square {
    pub fn new(file: u8, rank: u8) -> Square {
        Square { file, rank }
    }
}

impl fmt::Display for Square {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}{}", self.file, self.rank)
    }
}

////////////////////////////////////////////////////////////////////////////////

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum PieceType {
    Pawn,
    Lance,
    Knight,
    Silver,
    Gold,
    Bishop,
    Rook,
    King,
    ProPawn,
    ProLance,
    ProKnight,
    ProSilver,
    Horse,
    Dragon,
    All,
}

impl fmt::Display for PieceType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let pt = match *self {
            PieceType::Pawn => "FU",
            PieceType::Lance => "KY",
            PieceType::Knight => "KE",
            PieceType::Silver => "GI",
            PieceType::Gold => "KI",
            PieceType::Bishop => "KA",
            PieceType::Rook => "HI",
            PieceType::King => "OU",
            PieceType::ProPawn => "TO",
            PieceType::ProLance => "NY",
            PieceType::ProKnight => "NK",
            PieceType::ProSilver => "NG",
            PieceType::Horse => "UM",
            PieceType::Dragon => "RY",
            PieceType::All => "AL",
        };
        write!(f, "{}", pt)
    }
}

////////////////////////////////////////////////////////////////////////////////

type Board = [[Option<(Color, PieceType)>; 9]; 9];

#[derive(Debug, Default, PartialEq, Eq)]
pub struct Position {
    pub drop_pieces: Vec<(Square, PieceType)>,
    pub bulk: Option<Board>,
    pub add_pieces: Vec<(Color, Square, PieceType)>,
    pub side_to_move: Color,
}

impl fmt::Display for Position {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        if let Some(ref bulk) = self.bulk {
            for (i, row) in bulk.iter().enumerate() {
                write!(f, "P{}", i + 1)?;

                for pc in row.iter() {
                    match *pc {
                        Some((ref color, ref pt)) => write!(f, "{}{}", color, pt)?,
                        None => write!(f, " * ")?,
                    }
                }

                writeln!(f)?;
            }
        } else {
            write!(f, "PI")?;
            for &(ref sq, ref pt) in &self.drop_pieces {
                write!(f, "{}{}", sq, pt)?;
            }
            writeln!(f)?;
        }

        for &(ref color, ref sq, ref pt) in &self.add_pieces {
            writeln!(f, "P{}{}{}", color, sq, pt)?;
        }

        writeln!(f, "{}", self.side_to_move)?;

        Ok(())
    }
}

////////////////////////////////////////////////////////////////////////////////

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum Action {
    Move(Color, Square, Square, PieceType),
    Toryo,
    Chudan,
    Sennichite,
    TimeUp,
    IllegalMove,
    IllegalAction(Color),
    Jishogi,
    Kachi,
    Hikiwake,
    Matta,
    Tsumi,
    Fuzumi,
    Error,
}

impl fmt::Display for Action {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            Action::Move(ref color, ref from, ref to, ref pt) => {
                write!(f, "{}{}{}{}", color, from, to, pt)
            }
            Action::Toryo => write!(f, "%TORYO"),
            Action::Chudan => write!(f, "%CHUDAN"),
            Action::Sennichite => write!(f, "%SENNICHITE"),
            Action::TimeUp => write!(f, "%TIME_UP"),
            Action::IllegalMove => write!(f, "%ILLEGAL_MOVE"),
            Action::IllegalAction(ref color) => write!(f, "%{}ILLEGAL_ACTION", color),
            Action::Jishogi => write!(f, "%JISHOGI"),
            Action::Kachi => write!(f, "%KACHI"),
            Action::Hikiwake => write!(f, "%HIKIWAKE"),
            Action::Matta => write!(f, "%MATTA"),
            Action::Tsumi => write!(f, "%TSUMI"),
            Action::Fuzumi => write!(f, "%FUZUMI"),
            Action::Error => write!(f, "%ERROR"),
        }
    }
}

////////////////////////////////////////////////////////////////////////////////

#[derive(Debug, PartialEq, Eq)]
pub struct MoveRecord {
    pub action: Action,
    pub time: Option<Duration>,
}

impl fmt::Display for MoveRecord {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "{}", self.action)?;

        if let Some(ref time) = self.time {
            writeln!(f, "T{}", time.as_secs())?;
        }

        Ok(())
    }
}

////////////////////////////////////////////////////////////////////////////////

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn piece_type() {
        assert_eq!(&PieceType::Pawn.to_string(), "FU");
        assert_eq!(&PieceType::Lance.to_string(), "KY");
        assert_eq!(&PieceType::Knight.to_string(), "KE");
        assert_eq!(&PieceType::Silver.to_string(), "GI");
        assert_eq!(&PieceType::Gold.to_string(), "KI");
        assert_eq!(&PieceType::Bishop.to_string(), "KA");
        assert_eq!(&PieceType::Rook.to_string(), "HI");
        assert_eq!(&PieceType::King.to_string(), "OU");
        assert_eq!(&PieceType::ProPawn.to_string(), "TO");
        assert_eq!(&PieceType::ProLance.to_string(), "NY");
        assert_eq!(&PieceType::ProKnight.to_string(), "NK");
        assert_eq!(&PieceType::ProSilver.to_string(), "NG");
        assert_eq!(&PieceType::Horse.to_string(), "UM");
        assert_eq!(&PieceType::Dragon.to_string(), "RY");
        assert_eq!(&PieceType::All.to_string(), "AL");
    }

    #[test]
    fn action() {
        assert_eq!(
            &Action::Move(
                Color::Black,
                Square::new(7, 7),
                Square::new(7, 6),
                PieceType::Pawn,
            )
            .to_string(),
            "+7776FU"
        );

        assert_eq!(&Action::Toryo.to_string(), "%TORYO");
        assert_eq!(&Action::Chudan.to_string(), "%CHUDAN");
        assert_eq!(&Action::Sennichite.to_string(), "%SENNICHITE");
        assert_eq!(&Action::TimeUp.to_string(), "%TIME_UP");
        assert_eq!(&Action::IllegalMove.to_string(), "%ILLEGAL_MOVE");
        assert_eq!(
            &Action::IllegalAction(Color::Black).to_string(),
            "%+ILLEGAL_ACTION"
        );
        assert_eq!(
            &Action::IllegalAction(Color::White).to_string(),
            "%-ILLEGAL_ACTION"
        );
        assert_eq!(&Action::Jishogi.to_string(), "%JISHOGI");
        assert_eq!(&Action::Kachi.to_string(), "%KACHI");
        assert_eq!(&Action::Hikiwake.to_string(), "%HIKIWAKE");
        assert_eq!(&Action::Matta.to_string(), "%MATTA");
        assert_eq!(&Action::Tsumi.to_string(), "%TSUMI");
        assert_eq!(&Action::Fuzumi.to_string(), "%FUZUMI");
        assert_eq!(&Action::Error.to_string(), "%ERROR");
    }

    #[test]
    fn game_record() {
        let mut g = GameRecord {
            black_player: Some("NAKAHARA".to_string()),
            white_player: Some("YONENAGA".to_string()),
            event: Some("13th World Computer Shogi Championship".to_string()),
            site: Some("KAZUSA ARC".to_string()),
            start_time: Some(Time {
                date: time::Date::from_calendar_date(2003, time::Month::May, 3).unwrap(),
                time: Some(time::Time::from_hms(10, 30, 0).unwrap()),
            }),
            end_time: Some(Time {
                date: time::Date::from_calendar_date(2003, time::Month::May, 3).unwrap(),
                time: Some(time::Time::from_hms(11, 11, 5).unwrap()),
            }),
            time_limit: Some(TimeLimit {
                main_time: Duration::from_secs(1500),
                byoyomi: Duration::from_secs(0),
            }),
            opening: Some("YAGURA".to_string()),
            ..GameRecord::default()
        };
        g.moves.push(MoveRecord {
            action: Action::Move(
                Color::Black,
                Square::new(8, 7),
                Square::new(8, 6),
                PieceType::Pawn,
            ),
            time: Some(Duration::from_secs(5)),
        });
        g.moves.push(MoveRecord {
            action: Action::Toryo,
            time: None,
        });

        let csa = "\
V2.2
N+NAKAHARA
N-YONENAGA
$EVENT:13th World Computer Shogi Championship
$SITE:KAZUSA ARC
$START_TIME:2003/05/03 10:30:00
$END_TIME:2003/05/03 11:11:05
$TIME_LIMIT:00:25+00
$OPENING:YAGURA
PI
+
+8786FU
T5
%TORYO
";

        assert_eq!(csa, g.to_string());
    }
}
